runtime/freelist.c

```diff
@@ -1523,7 +1523,7 @@ static header_t *bf_allocate (mlsize_t wosz)
       return Hp_val (block);
     }else{
       /* allocate from the next available size */
-      mlsize_t s = ffs (bf_small_map & ((-1) << wosz));
+      mlsize_t s = ffs (bf_small_map & ((~0U) << wosz));
```

Comment:
`memory.c` has the following macro:

```
#define Page_mask ((uintnat) -1 << Page_log)
```

Is there a reason why `freelist.c` results in a warning but not that code in `memory.c`? (Is it because `Page_log` is a known constant?)

Reply:
The cast has greater priority over the shift. So `-1` is cast to unsigned int, has all its bits set to one and there’s no undefined behaviour.
I think this variant is more explicit in the intended meaning: `~(uintnat)0`. I’ve changed that in the same commit, but I can split the changes or not include it at all if that’s not needed.